src/storage-client/src/source/persist_source.rs

```diff
@@ -528,18 +588,270 @@ impl PersistSourceDataStats<'_> {
+                    for (mut part_time, d) in parts.drain(..) {
+                        part_time.1 = part_number;
+                        part_number += 1;
+                        yield Either::Right((part_time, d))
+                    }
+                    break;
+                }
+                Some(Event::Data(cap, data)) => {
+                    for d in data.drain(..) {
+                        parts.push((cap.time().clone(), d));
+                    }
+                }
+                Some(Event::Progress(prog)) => {
+                    let mut i = 0;
+                    parts.sort_by_key(|val| val.0.clone());
+                    // This can be replaced with `Vec::drain_filter` when it stabilizes.
+                    // `drain_filter_swapping` doesn't work as it reorders the vec.
+                    while i < parts.len() {
+                        if !prog.less_equal(&parts[i].0) {
+                            let (mut part_time, d) = parts.remove(i);
+                            part_time.1 = part_number;
+                            part_number += 1;
+                            yield Either::Right((part_time, d))
+                        } else {
+                            i += 1;
+                        }
+                    }
+                    yield Either::Left(prog)
+                }
+            }
+        }
+    });
+    let shutdown_button = builder.build(move |caps| async move {
+        // The output capability.
+        let mut cap_set = CapabilitySet::from_elem(caps.into_element());
+
+        // The frontier of our output. This matches the `CapabilitySet` above.
+        let mut output_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        // The frontier of the `flow_control` input.
+        let mut flow_control_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // Parts we have emitted, but have not yet retired (based on the `flow_control` edge).
+        let mut inflight_parts = Vec::new();
+        // Parts we have not yet emitted, but do participate in the `input_frontier`.
+        let mut pending_parts = std::collections::VecDeque::new();
+
+        // Only one worker is responsible for distributing parts
+        if worker_index != chosen_worker {
+            trace!(
+                "We are not the chosen worker ({}), exiting...",
+                chosen_worker
+            );
+            return;
+        }
+        tokio::pin!(data_input);
+        'emitting_parts: loop {
+            // At the beginning of our main loop, we determine the total size of
+            // inflight parts.
+            let inflight_bytes: usize = inflight_parts.iter().map(|(_, size)| size).sum();
+
+            // There are 2 main cases where we can continue to emit parts:
+            // - The total emitted bytes is less than `flow_control_max_bytes`.
+            // - We have not yet reached the `flow_control_frontier`
```

Comment:
Can we use not beyond terminology here?
```suggestion
            // - The output frontier is not beyond the `flow_control_frontier`
```